### PR TITLE
Optional real time profiling support.

### PR DESCRIPTION
**Pull request by bawr**
TL;DR: `time.sleep()` or waiting on IO can be detected now.

Rationale: vmprof is an excellent low-overhead profiler, but sometimes what we really want is to measure real time usage of the process, to detect situations when we're waiting for something else to happen.

That's especially handy when it comes to optimizing your code at a high-level, when the bottleneck is not in the Python code itself, but in the way it interacts with the rest of the system. Similarly, for that use case we want to sample all the threads, not just the thread that's currently running.

(Exact analysis of thread behaviour can be done downstream, because vmprof is already saving the thread ID for every stack snapshot.)

I'd love to support OSX, but I don't have a machine to test its signal behaviour.

Cheers!

### CODE REVIEW

File: src/vmprof_main.h [hunk 1]
```diff
@@ -175,11 +179,43 @@ static PY_THREAD_STATE_T * _get_pystate_for_this_thread(void) {
 }
 #endif
 
+#ifdef VMPROF_LINUX
+static void broadcast_signal_for_threads(pid_t pid)
```
Comment (planrich) [inline src/vmprof_main.h]: I do not see opendir in the list of signal safe functions. Neither is readdir_r and atoi. I might be wrong, but we need to double check if they are signal safe! Maybe there is another API that enumerates the thread ids of the process without asking the disk!

File: src/vmprof_main.h [hunk 2]
```diff
@@ -207,6 +243,17 @@ static void sigprof_handler(int sig_nr, siginfo_t* info, void *ucontext)
...
 #endif
 
+#ifdef VMPROF_LINUX
+    // SIGNAL ABUSE AHEAD
+    // On linux, the prof timer will deliver the signal to the thread which triggered the timer,
+    // because these timers are based on process and system time, and as such, are thread-aware.
+    // For the real timer, the signal gets delivered to the main thread, seemingly always.
+    // Consequently if we want to sample all threads, we need to forward the signal with tgkill.
+    if ((signal_type == SIGALRM) && (pid == tid)) {
+        broadcast_signal_for_threads(pid);
```
Comment (planrich) [inline src/vmprof_main.h]: I'm fine with this hack, I need to test it a little though. One question I have:

1) signal A (to main thread tid == pid == 1) arrives and uses tgkill to send a new signal to thread 2-3... (assuming there are 3 threads (1-3))
2) signal B to thread 2... OK
3) signal C to thread 1 (issuing more signals to thread 2 and 3)
4) signal D  to thread 3 ...

Will thread 3 get another signal initiated by signal C? Is it swallowed?
If the signal dispatching takes to long I think that should be recoginized...

File: src/vmprof.h [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #define PROFILE_LINES  '\x02'
 #define PROFILE_NATIVE '\x04'
 #define PROFILE_RPYTHON '\x08'
+#define PROFILE_REAL_TIME '\x0f'
```
Comment (bawr) [inline src/vmprof.h]: D'oh! Yes, this is wrong, and should be `'\x10'`.

File: src/vmprof_main.h (hunk 1 above)
Comment (bawr) [inline src/vmprof_main.h]: Ah, good call. In turn:
1. What can we do if they're *not* signal safe?
2. I've actually tried to find another API for thread enumeration and failed, I don't like the readdir solution, either. (Although technically it's asking the kernel, not the disk.)

File: src/vmprof_main.h (hunk 2 above)
Comment (bawr) [inline src/vmprof_main.h]: I'm pretty sure these signals are allowed to be coalesced / swallowed, yes.
I don't think there's a good way to detect this, but I might be wrong?

Possible workarounds:
1. Ignore it, when we read the profile later we can check for lost signals / stack traces if we know the threads were present for the whole profile duration. If we don't know that, it gets tricky.
2. Trigger a real-time signal from the main thread instead, and set other threads to catch that real-time signal instead of the SIGALRM from the timer?
3. Other?

File: src/vmprof_main.h (hunk 1 above)
Comment (planrich) [inline src/vmprof_main.h]: Yes, true the kernel delivers that info. If they calls are not signal safe, the program most likely freezes or crashes.
We could think about hooking into thread creation, that would register/unregister the thread id. Though a solution asking the kernel would be much better.

**Comment by bawr** [comment]
Hmm. Instead of waiting for signals and re-broadcasting them from the signal handler, can we instead spawn a designated signalling thread, whose only function is to act as a timer and deliver the signals to other threads in the process?

Then we could use thread-unsafe functions, but we'd need to handle forks correctly, spawning another signalling thread per-process, I guess.

**Comment by bawr** [comment]
Here's a simplified verion that's maybe more suitable to start with - we pin ourselves to the thread that requested the profile, and make sure all profiling is done on that thread, ignoring others. A better option for thread handling could be the next version of the feature, maybe?

Ideally, in the fullness of time, we'd go from supporting one real-time thread, to multiple threads, to multiple threads and/or processes, but it's useful to start small. I also added some simple tests to show that it's working, and how.

**Comment by bawr** [comment]
(Note: even with obvious small fixes, this still fails the tests on RPython, I'll probably need a hand there.)

**Comment by bawr** [comment]
Aaaand here's a version where we perform no signal-unsafe functions in the handler.

We can register and unregister threads from sampling, and in real time mode it's up to the caller.

**Comment by bawr** [comment]
Here's a small patch for PyPy, now all the tests pass over there, too.
https://gist.github.com/bawr/d80d99278b3619a6646c80646fa0dcf5

**Comment by bawr** [comment]
Any progress on this?

I'll be free to do some open-source stuff within the next week, so if I can make some changes to push it forward, I'll do just that. ;)

**Comment by planrich** [comment]
Looks good to me. Though the osx support is missing right? I'll merge it if we both support linux & osx.

**Comment by bawr** [comment]
Seems to work on OSX now. :tada:

File: src/vmprof_main.h [hunk 4]
```diff
@@ -181,11 +184,48 @@ static PY_THREAD_STATE_T * _get_pystate_for_this_thread(void) {
...
+#ifdef VMPROF_UNIX
+static int broadcast_signal_for_threads(void)
+{
+    int done = 1;
+    size_t i = 0;
+    pthread_t self = pthread_self();
+    pthread_t tid;
+    while (i < thread_count) {
+        tid = threads[i];
+        if (pthread_equal(tid, self))
```
Comment (planrich) [inline src/vmprof_main.h]: Can we add curly brackets for if & else here? it is really hard to read and potential risk for errors

File: src/vmprof_main.h [hunk 5]
```diff
@@ -199,6 +239,21 @@ static void sigprof_handler(int sig_nr, siginfo_t* info, void *ucontext)
...
+
+#ifdef VMPROF_UNIX
+    // SIGNAL ABUSE AHEAD
+    // On linux, the prof timer will deliver the signal to the thread which triggered the timer,
+    // because these timers are based on process and system time, and as such, are thread-aware.
+    // For the real timer, the signal gets delivered to the main thread, seemingly always.
+    // Consequently if we want to sample multiple threads, we need to forward this signal.
+    if (signal_type == SIGALRM) {
+        if (is_main_thread() && broadcast_signal_for_threads()) {
+            __sync_lock_release(&spinlock);
```
Comment (planrich) [inline src/vmprof_main.h]: Unsure why the spinlock is released here. It occurs to me as if the spinlock is only used to protect the threads array?

Comment (bawr) [inline src/vmprof_main.h]: Yes and no - it's used to protect whatever it was normally protecting *and* the threads array.

If we don't need to sample the main thread (when it's not registered for real-time sampling in the first place), we can bail out early here, because we just need to broadcast the signal - but we still need to release the spinlock before the immediate return.

File: src/vmprof_main.h (hunk 4 above)
Comment (bawr) [inline src/vmprof_main.h]: Done!

**Comment by bawr** [comment]
Rebased against upstream master and addressed the comments. Cheers! ;)

**Comment by planrich** [comment]
Thanks again!

**Comment by bawr** [comment]
Cool! Could you maybe make a .dev version with this available on pip?

I'd like to test some stuff about it later, and it's much simpler to set up if I can just pip install. ;)

**Comment by planrich** [comment]
Yes I did, there is 0.4.6.dev0

**Comment by planrich** [comment]
I'm currently rewriting some C code in Python. However I noticed that real time profiling is now some times not working. see: https://travis-ci.org/vmprof/vmprof-python/jobs/239723456

Could you maybe have a look at that? What is the reason for that?

**Comment by bawr** [comment]
Interesting... I can have a proper look over the weekend, will ping you earlier if it's something obvious.

**Comment by planrich** [comment]
Thanks. Currently I disabled one test with xfail...
